### PR TITLE
Add SAMD51 HAL

### LINKED ISSUE #38: ATSAMD51 support
Hello,

I've seen some mentions you're working on ATSAMD51 support:

- https://github.com/atsamd-rs/atsamd/issues/19
- https://github.com/rust-embedded/wg/issues/286

I'm interested in working on a Rust project targeting ATSAMD51, and was curious what the present state of that looks like.

### PR DESCRIPTION
**Pull request by sajattack**
Huge PR sorry. Some of the api had to change a bit just due to the differences in which peripherals need to be passed in for clock initialization. It's been so long I can't remember everything I changed but I got the blinky working and I'm so happy.

Fixes #38

### CODE REVIEW

**Comment by sajattack** [comment]
gif

**Comment by sajattack** [comment]
I'm testing UART now and it doesn't seem to be working.

**Comment by wez** [comment]
Nice!  I'm a combination of traveling and sick (great combo over the holidays :-/) so I'm not sure how much focused time I'll have to review this over the next week or so, but I am looking forward to it!

**Comment by sajattack** [comment]
Ok, take your time. The problem with my UART seems to be in the baud register calculation. My scale factor is ending up as -16220875724 which is not a u64.

**Comment by sajattack** [comment]
@wez I think this is finally ready.

**Comment by sajattack** [comment]
image

**Comment by sajattack** [comment]
[Happycat](https://i.imgur.com/Y6QYavk.jpg)

File: hal/src/clock51.rs [hunk 1]
```diff
@@ -0,0 +1,491 @@
+//! Configuring the system clock sources.
+//! You will typically need to create an instance of `GenericClockController`
+//! before you can set up most of the peripherals on the atsamd21 device.
```
Comment (djmcgill) [inline hal/src/clock51.rs]:
```suggestion
//! before you can set up most of the peripherals on the atsamd51 device.
```
Very minor typo

**Pull request merged**